### ISSUE TITLE
AssociativeSequenceKeys in `kyaml/yaml/types.go` causes inadvertent deletion of list items with merge3

### ISSUE THREAD

**Issue by khrisrichardson** [kubernetes-sigs/kustomize#2148]
I've been dabbling with the hugely beneficial alpha `config merge3` command and have a few observations.

There appear to be a couple keys missing from `AssociativeSequenceKeys`: namely `key`, `port`, and `uid`.

```bash
curl -LSfs http://localhost:8001/openapi/v2 |
  jq '[..
      | if type == "object" and
          .["x-kubernetes-patch-strategy"] == "merge"
        then
          .["x-kubernetes-patch-merge-key"] // empty
        else
          empty
        end
      ]
      | unique'
[
  "containerPort",
  "devicePath",
  "ip",
  "key",
  "mountPath",
  "name",
  "port",
  "topologyKey",
  "type",
  "uid"
]
```

And the order by which the keys are evaluated is somewhat problematic. I noticed this when the `additionalPrinterColumns` property of type `io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceColumnDefinition` array had all the entries with the same type (`string`) replaced with duplicates, the reason being that `type` takes precedence over `name` in the `AssociativeSequenceKeys`  string array.

It would probably be preferable to determine the patch merge key from the openapi spec, but in lieu of that I would propose the following modification, but am still gathering evidence:

```diff
diff --git a/kyaml/yaml/types.go b/kyaml/yaml/types.go
index 42c41961..31536d16 100644
--- a/kyaml/yaml/types.go
+++ b/kyaml/yaml/types.go
@@ -615,7 +615,7 @@ func (rn *RNode) VisitElements(fn func(node *RNode) error) error {
 // in Resources in a list, then the FIRST key which ALL elements in the list have is used as the
 // associative key for merging that list.
 var AssociativeSequenceKeys = []string{
-       "mountPath", "devicePath", "ip", "type", "topologyKey", "name", "containerPort",
+       "containerPort", "devicePath", "key", "mountPath", "port", "uid", "name", "type", "ip", "topologyKey",
 }

 // IsAssociative returns true if all elements in the list contain an AssociativeSequenceKey
```

Note that `containerPort` and `name` are in the same list of properties for `io.k8s.api.core.v1.ContainerPort`, but with `containerPort` being the patch merge key it should take precedence over `name`.

**Comment by khrisrichardson** [comment]
Just for a quick visual, here are the current definitions with multiple keys to be evaluated:

```bash
curl -LSfs http://localhost:8001/openapi/v2 |
  jq '(
    [
     (
      [..
      |  if type == "object" and
           .["x-kubernetes-patch-strategy"] == "merge"
         then
           .["x-kubernetes-patch-merge-key"] // empty
         else
           empty
        end
      ]
      | unique
     ) as $patchMergeKeys
     |
     [.definitions
     | to_entries
     |.[]
     | select(.key
             | startswith("io.k8s")
             )
     |{
       definition: .key,
       properties: [
         .value
          .properties? // {}
        | keys[]
        | . as $key
        | select(any($patchMergeKeys[]; . == $key))
       ]
      }
     ]
     |.[]
     | select(.properties
             | length > 1
             )
    ]
  )' |
  underscore --wrapwidth 100 pretty
[
  {
    definition: "io.k8s.api.admissionregistration.v1.ServiceReference",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.api.admissionregistration.v1beta1.ServiceReference",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.api.core.v1.AttachedVolume",
    properties: ["devicePath", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.ConfigMapKeySelector",
    properties: ["key", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.ConfigMapNodeConfigSource",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.api.core.v1.ContainerPort",
    properties: ["containerPort", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.EndpointPort",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.api.core.v1.ObjectReference",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.api.core.v1.SecretKeySelector",
    properties: ["key", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.ServicePort",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.api.core.v1.VolumeDevice",
    properties: ["devicePath", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.VolumeMount",
    properties: ["mountPath", "name"]
  },
  {
    definition: "io.k8s.api.discovery.v1beta1.EndpointPort",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceColumnDefinition",
    properties: ["name", "type"]
  },
  {
    definition: "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.ServiceReference",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceColumnDefinition",
    properties: ["name", "type"]
  },
  {
    definition: "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.ServiceReference",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta_v2",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.OwnerReference",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.OwnerReference_v2",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.StatusDetails",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.StatusDetails_v2",
    properties: ["name", "uid"]
  },
  {
    definition: "io.k8s.kube-aggregator.pkg.apis.apiregistration.v1.ServiceReference",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.kube-aggregator.pkg.apis.apiregistration.v1beta1.ServiceReference",
    properties: ["name", "port"]
  }
]
```

**Comment by khrisrichardson** [comment]
From the list of objects above with potentially conflicting merge keys as per `AssociativeSequenceKeys`, if one were to omit objects lacking a strategic merge key, one would be left with:

```json
[
  {
    definition: "io.k8s.api.core.v1.ContainerPort",
    patchMergeKey: "containerPort",
    properties: ["containerPort", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.ServicePort",
    patchMergeKey: "port",
    properties: ["name", "port"]
  },
  {
    definition: "io.k8s.api.core.v1.VolumeDevice",
    patchMergeKey: "devicePath",
    properties: ["devicePath", "name"]
  },
  {
    definition: "io.k8s.api.core.v1.VolumeMount",
    patchMergeKey: "mountPath",
    properties: ["mountPath", "name"]
  },
  {
    definition: "io.k8s.apimachinery.pkg.apis.meta.v1.OwnerReference",
    patchMergeKey: "uid",
    properties: ["name", "uid"]
  }
]
```

... which points to something more nefarious at work. By defining `AssociativeSequenceKeys` as universally applicable without referencing the openapi spec, keys are being considered as merge keys just by name alone even when they are not.

Considering this test case which I was going to add to `kyaml/yaml/merg3/list_test.go`, I don't think there's actually anything prohibiting the same name from appearing twice (perhaps the combination of fields needs to be unique), but based on the existing behavior of `AssociativeSequenceKeys`, fields with duplicates values for the "merge" keys are squashed.

```go
        {`Add list item with potentially conflicting merge keys: (name, type)`,
`
apiVersion: apiextensions/v1
spec:
  additionalPrinterColumns:
  - name: Age
    type: date
  - name: Ready
    type: string
`, `
apiVersion: apiextensions/v1
spec:
  additionalPrinterColumns:
  - name: Age
    type: date
  - name: Ready
    type: string
  - name: Reason
    type: string
`, `
apiVersion: apiextensions/v1
spec:
  additionalPrinterColumns:
  - name: Age
    type: date
  - name: Ready
    type: string
`, `
apiVersion: apiextensions/v1
spec:
  additionalPrinterColumns:
  - name: Age
    type: date
  - name: Ready
    type: string
  - name: Reason
    type: string
`, nil},
}
```

@pwittrock I believe you were doing some work around referencing the openapi spec. Do you have any pointers on how to address the trampling of values due to the improper handling of merge keys?

**Comment by pwittrock** [comment]
@khrisrichardson we could switch merge3 to use the openAPI -- it is there now and would be pretty simple to use instead of the hard-coded list.  We may even have individual fields annotate themselves with their merge key using openAPI embedded in the config as a comment.

Thoughts?

**Comment by khrisrichardson** [comment]
@pwittrock I've seen the recent work you've done WRT openAPI, but have yet to fully grasp the potential of the embedded comments. Let me see if I can make sense of the recent changes and swap out the hard-coded list. v4 looks to be shaping up to be a helluva makeover.

**Comment by pwittrock** [comment]
@khrisrichardson has this been resolved / can we close this issue?

**Comment by khrisrichardson** [comment]
Oh shoot, @pwittrock. Yes, it's resolved in #2245. Thanks, and great job on the release of kpt, the existence of which will greatly improve my workflow, and remove any need I had to use kyaml as a library in the first place.
